src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/DownloadManager.java

```diff
@@ -194,24 +217,60 @@ public Path download(
       }
     }
 
-    try {
-      downloader.download(
-          urls, authHeaders, checksum, canonicalId, destination, eventHandler, clientEnv, type);
-    } catch (InterruptedIOException e) {
-      throw new InterruptedException(e.getMessage());
+    if (disableDownload) {
+      throw new IOException(
+          String.format("Failed to download repo %s: download is disabled.", repo));
+    }
+
+    if (rewrittenUrls.isEmpty() && !originalUrls.isEmpty()) {
+      throw new IOException(getRewriterBlockedAllUrlsMessage(originalUrls));
+    }
+
+    for (int attempt = 0; attempt <= retries; ++attempt) {
+      try {
+        downloader.download(
+            rewrittenUrls,
+            rewrittenAuthHeaders,
+            checksum,
+            canonicalId,
+            destination,
+            eventHandler,
+            clientEnv,
+            type);
+        break;
+      } catch (ContentLengthMismatchException e) {
+        if (attempt == retries) {
+          throw e;
+        }
+      } catch (InterruptedIOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)